bluepysnap/node_sets.py

```diff
@@ -0,0 +1,99 @@
+# Copyright (c) 2020, EPFL/Blue Brain Project
+
+# This file is part of BlueBrain SNAP library <https://github.com/BlueBrain/snap>
+
+# This library is free software; you can redistribute it and/or modify it under
+# the terms of the GNU Lesser General Public License version 3.0 as published
+# by the Free Software Foundation.
+
+# This library is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+# details.
+
+# You should have received a copy of the GNU Lesser General Public License
+# along with this library; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+"""Access to node set data."""
+
+import collections
+from copy import deepcopy
+
+import numpy as np
+from bluepysnap.exceptions import BluepySnapError
+from bluepysnap import utils
+
+
+def _sanitize(node_set):
+    for key, values in node_set.items():
+        if isinstance(values, list):
+            if len(values) == 1:
+                node_set[key] = values[0]
+            else:
+                # sorted unique value list
+                node_set[key] = np.unique(np.asarray(values)).tolist()
+    return node_set
+
+
+def _resolve_set(content, resolved, set_name):
+    """Resolve a node set from content.
+
+    Notes:
+        if the node set is a combined node set then the function resolve all the node sets
+        included in the combined one.
+    """
+    if set_name in resolved:
+        # return already resolved node_sets
+        return resolved[set_name]
+
+    set_value = content.get(set_name)
+    if set_value is None:
+        raise BluepySnapError("Missing node_set: '{}'".format(set_name))
+    if not isinstance(set_value, (collections.Mapping, list)) or not set_value:
+        raise BluepySnapError("Ambiguous node_set: '{}'".format({set_name: set_value}))
+
+    # keep the content intact
+    set_value = deepcopy(set_value)
+
+    if isinstance(set_value, collections.Mapping):
+        return _sanitize(set_value)
```

Comment:
What about always adding to `resolved` what you are about to return ?
This way, you are sure that `resolved` is fully resolved when exiting the function. Otherwise, you still have to update it after it has run, as you do line 74 or line 65.

Reply:
indeed I changed it